rfcs/2023-09-12-second-relays.md

```diff
@@ -0,0 +1,121 @@
+
+2. Thin clients with hosted accounts.
+
+Host the current client on the server and have a thin client running on the devices
+
+Pros:
+- substantially reduces the traffic for mobile clients
+- probably the most attractive option for mass market users who expect to have an account on the server and be able to access it from multiple devices and via web.
+
+Cons:
+- a lot of design and development
+- makes accounts visible to the server
+- additional protocol design to protect connections graph
+
+Overall, this is not a viable option for the current stage.
+
+3. First step SMP and XFTP relays.
+
+Introduce SMP and XFTP protocol extenstions to allow message senders and file recipients to delegate the tasks of sending messages and receiving files to the relays, so that peer-chosen relays can only observe IP addresses of other relays and not of the users.
+
+Pros:
+- no dependency on Tor
+- lower latency
+- reduces client traffic, both due to retries handled by these additional relays and due to the smaller number of connections to the peer-chosen relays.
+- higher transport privacy: protects IP addresses from peers and makes transport correlation harder.
+- additional functions (sending pending and scheduled messages).
+- prevents session-based correlation of the traffic by the destination relay (assuming no collusion with the first-step relay), as there will be one session between first-step relay and destination relay mixing traffic from multiple users.
+
+Cons:
+- substantial design and development
+- can undermine delivery stability
+
+This seems an attractive option, both from technical (reasonable complexity) and positioning (moving it closer to nix networks, while still avoiding server list visibility), and it is the middle between using Tor, which would make the product more niche and has downsides and development cost too, and thin clients, which is much more work and creates a different threat model, that is not very attractive for the current stage.
+
+Below considers this design.
+
+# First step user-chosen relays
+
+## Design requirements
+
+1. To avoid increasing the complexity of self-hosting, we should not create additional server types, and existing SMP and XFTP servers should be extended to provide the 1st step relay functions.
+
+2. SMP sending relay should not be able to observe queue addresses and their count on the destination relays.
+
+3. There must be no identifiers and cyphertext in common in outgoing and incoming traffic in 1st step relays.
+
+4. Traffic between the client and destination relays must be e2e encrypted, with MITM mitigated using server identity (certificate fingerprint), ideally without adding the additional fingerprint in destination relay address.
+
+5. SMP sending relays should implement retry logic and hold messages while they are delivered. They also should push "sent" confirmation to the client. Question: does it mean the third tick? another kind of the first tick (or some other symbol) in the UX?
```

Comment:
third tick may be an overkill (there's even still possibility for "read" marker - too many unclear indications on a sent message), agent should probably report SENT to client after sending to the first relay, or maybe it could be a different status / event for debugging, for example SENT_RELAY